File: Java/Ema/PerfTools/src/main/java/com/refinitiv/ema/perftools/emajprovperf/IProviderThread.java
package com.refinitiv.ema.perftools.emajprovperf;

import com.refinitiv.ema.access.*;
import com.refinitiv.ema.perftools.common.ItemInfo;
import com.refinitiv.ema.perftools.common.ItemWatchList;
import com.refinitiv.ema.perftools.common.*;
import com.refinitiv.ema.rdm.EmaRdm;

import java.util.Map;
import java.util.*;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.locks.Condition;
import java.util.concurrent.locks.ReentrantLock;

import static com.refinitiv.ema.perftools.common.LogFileHelper.initFile;
import static com.refinitiv.ema.perftools.common.LogFileHelper.writeFile;

public class IProviderThread extends ProviderThread {

    private final IProviderPerfConfig config;
    private final ReentrantLock clientLock;
    private final Condition connectionCondition;

    private final LatencyRandomArray genMsgLatencyRandomArray;
    private final Map<Long, ItemInfo> itemHandles;

    private final ItemWatchList generics;
    private final Set<ItemInfo> closedItems;
    private final Set<Long> closedClients;

    private final GenericMsg genericMsg;

    private LogFileInfo latencyFile;
    private ProviderPerfClient client;
    private boolean connected;

    public IProviderThread(BaseProviderPerfConfig baseConfig, XmlMsgData msgData) {
        super(baseConfig, msgData);
        this.config = (IProviderPerfConfig) baseConfig;
        this.genMsgLatencyRandomArray = new LatencyRandomArray();
        this.itemHandles = new ConcurrentHashMap<>(MSG_STORE_CAPACITY);
        this.generics = new ItemWatchList(MSG_STORE_CAPACITY);
        this.closedItems = new HashSet<>(MSG_STORE_CAPACITY);
        this.closedClients = new HashSet<>(100);
        this.genericMsg = EmaFactory.createGenericMsg();
        this.clientLock = new ReentrantLock();
        this.connectionCondition = clientLock.newCondition();
    }

    @Override
    public void initialize(int threadIndex) {
        super.initialize(threadIndex);
        writeFile(statsFile, "UTC, Requests received, Images sent, Updates sent, Posts reflected, GenMsgs sent, GenMsg Latencies sent, GenMsgs received, GenMsg Latencies received, GenMsg Latency avg (usec), GenMsg Latency std dev (usec), GenMsg Latency max (usec), GenMsg Latency min (usec), CPU usage (%%), Memory (MB)\n");
        if (config.genMsgsPerSec() != 0 && config.latencyGenMsgRate() > 0) {
            randomArrayOptions.totalMsgsPerSec(config.genMsgsPerSec());
            randomArrayOptions.latencyMsgsPerSec(config.latencyGenMsgRate());
            randomArrayOptions.ticksPerSec(config.ticksPerSec());
            randomArrayOptions.arrayCount(LATENCY_RANDOM_ARRAY_SET_COUNT);

            if (genMsgLatencyRandomArray.create(randomArrayOptions) != PerfToolsReturnCodes.SUCCESS) {
                System.err.println("Error initializing application: Failed to create generic messages latency random array");
                System.exit(-1);
            }
        }

        // Open latency file if configured.
        if (config.logLatencyToFile()) {
            this.latencyFile = initFile(config.latencyFilename() + providerIndex + ".csv");
            writeFile(latencyFile, "Message type, Send time, Receive time, Latency (usec)\n");
        }
    }

    @Override
    protected void initializeOmmProvider() {
        this.client = new ProviderPerfClient(this.config, this.clientLock);
        final OmmIProviderConfig providerConfig = EmaFactory.createOmmIProviderConfig()
                .providerName(providerConfigName);
        if (this.config.useUserDispatch()) {
            providerConfig.operationModel(OmmIProviderConfig.OperationModel.USER_DISPATCH);
        }

        this.provider = EmaFactory.createOmmProvider(providerConfig, this.client); //Seems that we should create only one provider

        if (this.config.useUserDispatch()) {
            while (!connected) {
                try {
                    this.provider.dispatch(1000);
                } catch (OmmException e) {
                    shutdown = true;
                }
            }
        } else {
            try {
                clientLock.lock();
                while (!connected) {
                    try {
                        System.out.printf("ProviderThread_%d run. Waiting initilization... \n", this.providerIndex);
                        connectionCondition.await(1, TimeUnit.SECONDS);
                    } catch (InterruptedException e) {
                        shutdown = true;
                    }
                }
            } finally {
                clientLock.unlock();
            }
        }
    }

    @Override
    protected void executeMsgBurst(long nextTime) throws OmmInvalidUsageException {
        if (!this.updates.isEmpty()) {
            sendUpdateMessages();
        }
        if (!this.generics.isEmpty()) {
            sendGenericMessages();
        }
        if (!this.closedItems.isEmpty()) {
            processClosedHandles();
        }
        if (!this.closedClients.isEmpty()) {
            processClosedClients();
        }
        if (!this.refreshes.isEmpty()) {
            sendRefreshMessages();
        }
    }

    @Override
    protected void handleRefresh(ItemInfo itemInfo) {
        if ((itemInfo.itemFlags() & ItemFlags.IS_STREAMING_REQ) != 0) {
            updates.add(itemInfo);
            generics.add(itemInfo);
        }
    }

    @Override
    protected boolean sendMsg(Msg msg, ItemInfo itemInfo) {
        try {
            if (isActiveStream(itemInfo)) {
                submitMsg(msg, itemInfo.itemHandle());
                return true;
            }
        } catch (OmmInvalidUsageException e) {
            if (e.errorCode() == OmmInvalidUsageException.ErrorCode.INVALID_ARGUMENT) {
                closedClients.add(itemInfo.clientHandle());
                this.processClosedHandles();
                return false;
            }
            throw e;
        }
        return false;
    }

    private void sendGenericMessages() {
        //Determine generic messages to send out. Spread the remainder out over the first ticks
        int genMsgsLeft = config.genMsgsPerTick();
        int genMsgsPerTickRemainder = config.genMsgsPerTickRemainder();
        if (genMsgsPerTickRemainder > currentTicks) {
            ++genMsgsLeft;
        }

        int latencyGenMsgNumber = (config.latencyGenMsgRate() > 0) ? genMsgLatencyRandomArray.next() : -1;
        boolean setLatency = false;

        try {
            for (; genMsgsLeft > 0; --genMsgsLeft) {
                long latencyStartTime;
                ItemInfo itemInfo = generics.getNext();

                if (this.isActiveStream(itemInfo)) {
                    // When appropriate, provide a latency timestamp for the generic messages.
                    genericMsg.clear();
                    genericMsg.domainType(itemInfo.attributes().domainType());

                    switch (itemInfo.attributes().domainType()) {
                        case EmaRdm.MMT_MARKET_PRICE:
                            ItemData itemData = (ItemData) itemInfo.itemData();
                            int index = itemData.getNextIndex();
                            if (itemData.getNextIndex() == xmlMsgData.marketPriceUpdateMsgCount()) {
                                index = 0;
                            }
                            setLatency = config.latencyGenMsgRate() == BaseProviderPerfConfig.ALWAYS_SEND_LATENCY_GENMSG || latencyGenMsgNumber == (genMsgsLeft - 1);
                            genericMsg.payload(createMarketPricePayload(xmlMsgData.marketPriceGenMsgs()[index], TIM_TRK_3_FID, setLatency));
                            itemData.setNextIndex(++index);
                            break;
                        default:
                            return;
                    }

                    if (sendMsg(genericMsg, itemInfo)) {
                        if (setLatency) {
                            providerThreadStats.latencyGenMsgSentCount().increment();
                        }
                        providerThreadStats.genMsgSentCount().increment();
                    }
                }
            }
        } catch (OmmInvalidUsageException e) {
            if (e.errorCode() == OmmInvalidUsageException.ErrorCode.NO_BUFFERS) {
                this.providerThreadStats.outOfBuffersCount().add(genMsgsLeft);
                return;
            }
            throw e;
        }


        if (++currentTicks > this.config.ticksPerSec()) {
            currentTicks = 0;
        }
    }

    @Override
    protected void submitMsg(Msg msg, Long handle) {
        if (msg.dataType() == DataType.DataTypes.GENERIC_MSG) {
            this.provider.submit(genericMsg, handle);
        } else {
            super.submitMsg(msg, handle);
        }
    }

    private boolean isActiveStream(ItemInfo itemInfo) {
        try {
            clientLock.lock();
            return itemHandles.containsKey(itemInfo.itemHandle())
                    && itemHandles.get(itemInfo.itemHandle()).active()
                    && !closedClients.contains(itemInfo.clientHandle());
        } finally {
            clientLock.unlock();
        }
    }

    public void processClose(Long handle) {
        ItemInfo itemInfo = this.itemHandles.get(handle);
        if (Objects.nonNull(itemInfo) && itemInfo.active()) {
            closedItems.add(itemInfo);
            itemInfo.active(false);
        }
    }

    private void processClosedHandles() {
        try {
            clientLock.lock();
            this.itemHandles.entrySet().removeIf(e -> closedItems.contains(e.getValue()));
            this.refreshes.removeIf(closedItems::contains);
            this.generics.removeItems(this.closedItems);
            this.updates.removeItems(this.closedItems);
            this.itemInfoPool().returnToPool(this.closedItems);
            this.closedItems.clear();
        } finally {
            clientLock.unlock();
        }
    }

    private void processClosedClients() {
        try {
            clientLock.lock();
            refreshes.removeIf((e) -> closedClients.contains(e.clientHandle()));
            updates.removeItemsForClients(closedClients);
            generics.removeItemsForClients(closedClients);
            Iterator<Map.Entry<Long, ItemInfo>> entries = itemHandles.entrySet().iterator();
            while (entries.hasNext()) {
                Map.Entry<Long, ItemInfo> infoEntry = entries.next();
                if (closedClients.contains(infoEntry.getValue().clientHandle())) {
                    entries.remove();
                    itemInfoPool().returnToPool(infoEntry.getValue());
                }
            }
            closedClients.clear();
        } finally {
            clientLock.unlock();
        }
    }


    public void notifyOnConnected() {
        this.connected = true;
        if (clientLock.isLocked()) {
            connectionCondition.signalAll();
        }
    }

    public Map<Long, ItemInfo> itemHandles() {
        return itemHandles;
    }

    public Set<Long> closedClients() {
        return closedClients;
    }

    public LogFileInfo latencyFile() {
        return latencyFile;
    }

    @Override
    public void clear() {
        super.clear();
        if (latencyFile.supportedWriting()) {
            latencyFile.writer().close();
        }
    }
}

File: CSharp/Ema/PerfTools/EMA_IProvPerf/IProviderThread.cs
/*|-----------------------------------------------------------------------------
 *|            This source code is provided under the Apache 2.0 license
 *|  and is provided AS IS with no warranty or guarantee of fit for purpose.
 *|                See the project's LICENSE.md for details.
 *|           Copyright (C) 2024 LSEG. All rights reserved.     
 *|-----------------------------------------------------------------------------
 */

using LSEG.Ema.Access;
using LSEG.Ema.PerfTools.Common;
using LSEG.Ema.Rdm;
using System.Collections.Concurrent;
using static LSEG.Ema.Access.OmmProviderConfig;
using static LSEG.Ema.PerfTools.Common.ItemEncoder;

namespace LSEG.Ema.PerfTools.EMA_IProvPerf
{
    public class IProviderThread : ProviderThread
    {
        private EmaIProvPerfConfig config;
        private AutoResetEvent connectionCondition;

        private LatencyRandomArray genMsgLatencyRandomArray;
        internal ConcurrentDictionary<long, ItemInfo> itemHandles;

        private ItemWatchlist generics;
        internal HashSet<ItemInfo> closedItems;
        internal HashSet<long> closedClients;

        private GenericMsg genericMsg;

        private ProviderPerfClient? client;
        private bool connected;

        public override ProviderRoleEnum ProviderRoleKind => ProviderRoleEnum.INTERACTIVE;

        public IProviderThread(EmaIProvPerfConfig config, XmlMsgData msgData) : base(config, msgData)
        {
            this.config = config;
            genMsgLatencyRandomArray = new LatencyRandomArray();
            itemHandles = new ConcurrentDictionary<long, ItemInfo>(3, MSG_STORE_CAPACITY);
            generics = new ItemWatchlist(MSG_STORE_CAPACITY);
            closedItems = new HashSet<ItemInfo>(MSG_STORE_CAPACITY);
            closedClients = new HashSet<long>(100);
            genericMsg = new GenericMsg();
            connectionCondition = new(false);
        }

        public override void Initialize(int threadIndex)
        {
            base.Initialize(threadIndex);
            statsFile.WriteFile("UTC, Requests received, Images sent, Updates sent, Posts reflected, GenMsgs sent, GenMsg Latencies sent, GenMsgs received, GenMsg Latencies received, GenMsg Latency avg (usec), GenMsg Latency std dev (usec), GenMsg Latency max (usec), GenMsg Latency min (usec), CPU usage (%), Memory (MB)\n");
            if (config.genMsgsPerSec != 0 && config.latencyGenMsgRate > 0)
            {
                randomArrayOptions.TotalMsgsPerSec = config.genMsgsPerSec;
                randomArrayOptions.LatencyMsgsPerSec = config.latencyGenMsgRate;

                if (genMsgLatencyRandomArray.Create(randomArrayOptions) != PerfToolsReturnCode.SUCCESS)
                {
                    Console.Error.WriteLine("Error initializing application: Failed to create generic messages latency random array");
                    Environment.Exit(-1);
                }
            }

            // Open latency file if configured.
            if (config.LogLatencyToFile())
            {
                LatencyFile = LogFileHelper.InitFile(config.latencyFilename + providerIndex + ".csv");
                LatencyFile.WriteFile("Message type, Send time, Receive time, Latency (usec)\n");
            }
        }

        protected override void InitializeOmmProvider()
        {
            client = new ProviderPerfClient(config, this, clientLock);

            OmmIProviderConfig providerConfig = new OmmIProviderConfig()
                .ProviderName(providerConfigName!)
                .ServerCertificate(config.CertificateFilename!)
                .ServerPrivateKey(config.PrivateKeyFilename!);
            if (config.UseUserDispatch)
            {
                providerConfig.OperationModel(OmmIProviderConfig.OperationModelMode.USER_DISPATCH);
            }

            provider = new OmmProvider(providerConfig, this.client); //Seems that we should create only one provider

            try
            {
                if (config.UseUserDispatch)
                {
                    while (!(connected || shutdown))
                    {
                        provider.Dispatch(1000);
                    }
                }
                else
                {
                    var conditionIsSet = false;
                    while (!(conditionIsSet || shutdown))
                    {
                        conditionIsSet = connectionCondition.WaitOne(1000);
                    }
                }
            }
            catch (Exception)
            {
                shutdown = true;
            }
        }

        public override void ExecuteMsgBurst(long nextTime)
        {
            if (closedItems.Count != 0)
            {
                ProcessClosedHandles();
            }
            if (closedClients.Count != 0)
            {
                ProcessClosedClients();
            }
            if (updates.Count() != 0)
            {
                SendUpdateMessages();
            }
            if (generics.Count() != 0)
            {
                SendGenericMessages();
            }
            do
            {
                if (refreshes.Count != 0)
                {
                    SendRefreshMessages();
                }
                else
                {
                    break;
                }
            } while (GetTime.GetMicroseconds() < nextTime);
        }

        protected override void HandleRefresh(ItemInfo itemInfo)
        {
            if ((itemInfo.ItemFlags & (int)ItemFlags.IS_STREAMING_REQ) != 0)
            {
                updates.Add(itemInfo);
                generics.Add(itemInfo);
            }
        }

        protected override bool SendMsg(Msg msg, ItemInfo itemInfo)
        {
            try
            {
                if (IsActiveStream(itemInfo))
                {
                    SubmitMsg(msg, itemInfo);
                    return true;
                }
            }
            catch(OmmInvalidHandleException)
            {
                closedClients.Add(itemInfo.ClientHandle);
                ProcessClosedClients();
                return false;
            }
            catch (OmmInvalidUsageException e)
            {
                if (e.ErrorCode == OmmInvalidUsageException.ErrorCodes.INVALID_ARGUMENT)
                {
                    closedClients.Add(itemInfo.ClientHandle);
                    ProcessClosedClients();
                    return false;
                }
                throw;
            }
            return false;
        }

        internal void NotifyOnConnected()
        {
            connected = true;
            if (!config.UseUserDispatch)
            {
                connectionCondition.Set();
            }
        }

        internal ItemInfo CreateItemInfoFromMarketPriceRequest(RequestMsg reqMsg, long itemHandle, long clientHandle)
        {
            var flags = ItemFlags.IS_SOLICITED;
            if (reqMsg.InterestAfterRefresh())
            {
                flags |= ItemFlags.IS_STREAMING_REQ;
            }
            if (reqMsg.PrivateStream())
            {
                flags |= ItemFlags.IS_PRIVATE;
            }

            ItemInfo itemInfo = itemInfoPool.GetFromPool()!;
            itemInfo.ItemId = reqMsg.StreamId();
            itemInfo.ItemFlags = (int)flags;
            itemInfo.ClientHandle = clientHandle;
            itemInfo.ItemHandle = itemHandle;
            itemInfo.Active = true;
            itemInfo.Attributes.DomainType = reqMsg.DomainType();
            if (reqMsg.HasName && reqMsg.HasNameType)
            {
                itemInfo.Attributes.Name = reqMsg.Name();
                itemInfo.Attributes.NameType = reqMsg.NameType();
            }

            if (reqMsg.HasServiceId)
            {
                itemInfo.Attributes.ServiceId = reqMsg.ServiceId();
            }

            if (reqMsg.HasServiceName)
            {
                itemInfo.Attributes.ServiceName = reqMsg.ServiceName();
            }

            refreshes.Enqueue(itemInfo);

            return itemInfo;
        }

        private bool IsActiveStream(ItemInfo itemInfo)
        {
            Monitor.Enter(clientLock);
            try
            {
                return itemHandles.TryGetValue(itemInfo.ItemHandle, out var foundItemInfo)
                        && foundItemInfo.Active
                        && !closedItems.Contains(itemInfo)
                        && !closedClients.Contains(itemInfo.ClientHandle);
            }
            finally
            {
                Monitor.Exit(clientLock);
            }
        }

        internal void SendGenericMessages()
        {
            //Determine generic messages to send out. Spread the remainder out over the first ticks
            int genMsgsLeft = config.genMsgsPerTick;
            int genMsgsPerTickRemainder = config.genMsgsPerTickRemainder;
            if (genMsgsPerTickRemainder > currentTicks)
            {
                ++genMsgsLeft;
            }

            int latencyGenMsgNumber = (config.latencyGenMsgRate > 0) ? genMsgLatencyRandomArray.Next() : -1;
            bool setLatency = false;

            try
            {
                for (; genMsgsLeft > 0; --genMsgsLeft)
                {
                    ItemInfo? itemInfo = generics.GetNext();

                    if (itemInfo != null && IsActiveStream(itemInfo))
                    {
                        // When appropriate, provide a latency timestamp for the generic messages.
                        genericMsg.Clear();
                        genericMsg.DomainType(itemInfo.Attributes.DomainType);

                        switch (itemInfo.Attributes.DomainType)
                        {
                            case EmaRdm.MMT_MARKET_PRICE:
                                ItemData itemData = itemInfo.ItemData;
                                int index = itemData.NextIndex;
                                if (itemData.NextIndex == xmlMsgData.UpdateCount)
                                {
                                    index = 0;
                                }
                                setLatency = config.latencyGenMsgRate == BaseProviderPerfConfig.ALWAYS_SEND_LATENCY_GENMSG || latencyGenMsgNumber == (genMsgsLeft - 1);
                                var payload = CreateMarketPricePayload(xmlMsgData.MpGenMsgs![index], TIM_TRK_3_FID, setLatency);
                                if (payload != null)
                                {
                                    genericMsg.Payload(payload);
                                }
                                itemData.NextIndex = ++index;
                                break;
                            default:
                                return;
                        }

                        if (SendMsg(genericMsg, itemInfo))
                        {
                            if (setLatency)
                            {
                                ProviderThreadStats.LatencyGenMsgSentCount.Increment();
                            }
                            ProviderThreadStats.GenMsgSentCount.Increment();
                        }
                    }
                }
            }
            catch (OmmInvalidUsageException e)
            {
                if (e.ErrorCode == OmmInvalidUsageException.ErrorCodes.NO_BUFFERS)
                {
                    ProviderThreadStats.OutOfBuffersCount.Add(genMsgsLeft);
                    return;
                }
                throw;
            }


            if (++currentTicks > config.TicksPerSec)
            {
                currentTicks = 0;
            }
        }

        public void ProcessClose(long handle)
        {
            itemHandles.TryGetValue(handle, out ItemInfo? itemInfo);
            if (itemInfo != null && itemInfo.Active)
            {
                closedItems.Add(itemInfo);
                itemInfo.Active = false;
            }
        }

        private void ProcessClosedHandles()
        {
            Monitor.Enter(clientLock);
            try
            {
                foreach (var item in closedItems)
                {
                    itemHandles.Remove(item.ItemHandle, out _);
                    itemInfoPool.ReturnToPool(item);
                }

                refreshes.RemoveIf(closedItems.Contains);
                generics.RemoveIf(closedItems.Contains);
                updates.RemoveIf(closedItems.Contains);

                closedItems.Clear();
            }
            finally
            {
                Monitor.Exit(clientLock);
            }
        }

        private void ProcessClosedClients()
        {
            Monitor.Enter(clientLock);
            try
            {
                if(refreshes.Count > 0)
                    refreshes.RemoveIf(x => closedClients.Contains(x.ClientHandle));

                if(updates.Count() > 0)
                    updates.RemoveIf(x => closedClients.Contains(x.ClientHandle));

                if(generics.Count() > 0)
                    generics.RemoveIf(x => closedClients.Contains(x.ClientHandle));

                var itemHandleEntriesToRemove = itemHandles
                    .Where(x => closedClients.Contains(x.Value.ClientHandle))
                    .ToList();
                foreach (var itemEntry in itemHandleEntriesToRemove)
                {
                    itemHandles.Remove(itemEntry.Key, out _);
                    itemInfoPool.ReturnToPool(itemEntry.Value);
                }

                closedClients.Clear();
            }
            finally
            {
                Monitor.Exit(clientLock);
            }
        }
    }
}
